### ISSUE TITLE
Validation callback

### ISSUE THREAD

**Issue by guizmaii** [orgsync/active_interaction#258]
Hey,

Is there a callback when the validation fails ?
This kind of callback could be great because, thus, I can define a default behavior when validation fails and specific behaviors in controllers.

Jules

**Comment by tfausak** [comment]
There is no callback specifically for when validation fails, but you can create one easily.

``` rb
class Example < ActiveInteraction::Base
  string :string

  validates :string,
    presence: true

  set_callback :validate, :after,
    if: 'errors.present?' do
      # ...
    end
end
```

**Comment by guizmaii** [comment]
Does it work if I only have params validations, like this :
```ruby
string :string
```

??

I'm testing your solution but my callback is not called

**Comment by tfausak** [comment]
Ah, we call that type checking. It happens before validation. Check out #196 for why we make that distinction. Regardless, this should do what you want:

``` rb
class Example < ActiveInteraction::Base
  string :string

  set_callback :type_check, :after,
    if: 'errors.present?' do
      # ...
    end
end
```

**Comment by guizmaii** [comment]
Ok thanks. I was reading your (very clear) code and I just found your solution myself. 

Thanks !

And great job to add type checking to Ruby ! 
I love this functionality in your gem, I'm more a Java/Scala dev than a Ruby dev.

**Comment by guizmaii** [comment]
I don't know why but my callback is called twice.
I break only one of the type check rules defined in my interaction. 
In my callback, I do something like that : 

```ruby
errors.add(:data_validation_error, 'data_validation_error')
```

Add an error recall type checking ?

**Comment by guizmaii** [comment]
Ok, I think I found why it's call twice : 

When un call ```run``` method, it is called, as expected.
But when, in my controller, I ask if ```outcome.valid?``` it re-run the type checking and so replay my callback.

Is it normal ? I think it's not a great behavior.

**Comment by guizmaii** [comment]
Ok just for information, I put this crappy hack in my callback in order to execute my business logic only once : 

```ruby
if @callback_called_once
  errors.add(:data_validation_error, 'data_validation_error')
  ...
else
   @callback_called_once = true
end
```

So, with this, I'm sure that at the end, the ```:data_validation_error``` is in my errors.
At the first call of this callback, the instance variable will be nil and at the second call it will be true and will execute my business logic.

I debugged a little to understand what's happening and why this double call.
It's because the ```run``` method call the ```valid?``` method that will call the ```type_check``` method  via ```run_validation!``` method.
And when I call the ```valid?``` method in my controller it re-call the ```type_check``` via ```run_validation!``` method.

I think it a bug, what's your point of view about this ?

**Comment by tfausak** [comment]
I don't think this is a bug. This is just how ActiveModel works.

``` rb
require 'active_model' # 4.2.0

class ExampleModel
  include ActiveModel::Validations
  include ActiveModel::Validations::Callbacks

  after_validation { puts 'after validation' }
end

example_model = ExampleModel.new
example_model.valid?
# after validation
example_model.valid?
# after validation
```

**Comment by guizmaii** [comment]
I'm not agree with you and how ActiveModel works is not important here, I think.

The semantic of how we have to use ActiveInteractor with the use of the ```run``` method and then the use of ```outcome.valid?``` means that the ```run``` method will produce a result (the ```outcome```) and the ```valid?``` method will only check the produced result. Not re-run the validation the ```run``` method has already done.

If I speak in terms of types, the ```outcome``` should be a class like, for exemple, ```ActiveResult``` which contains a ```valid?``` method and contains accessible attributes that will contain the ```run``` method results. 

I terms of performance and semantic, it will be better, I think.

**Comment by tfausak** [comment]
How ActiveModel works is important here because ActiveInteraction behaves like an ActiveModel. (That's why we have [the `ActiveModelable` concern](https://github.com/orgsync/active_interaction/blob/v1.5.0/lib/active_interaction/concerns/active_modelable.rb#L7).) ActiveInteraction's integration with ActiveModel is one of its main value adds over other command pattern libraries like [Mutations](https://github.com/cypriss/mutations) or [Interactor](https://github.com/collectiveidea/interactor). 

I'm not sure exactly what you're trying to do, but generally I make validations idempotent. I don't assume they'll only be run once. If I want something to be run at most one time, I'll stick it in `execute` (or a callback around that). Here are some examples showing how validations and callbacks behave:

``` rb
# coding: utf-8

require 'active_interaction' # 1.5.0

class Example < ActiveInteraction::Base
  set_callback :type_check, :before do
    puts 'before type check'
  end

  string :string

  set_callback :validate, :before do
    puts 'before validate'
  end

  validates :string,
    presence: true

  set_callback :execute, :before do
    puts 'before execute'
  end

  def execute
    errors.add_sym :string unless string == 'valid'
  end
end

outcome = Example.run(string: nil)
# before type check
outcome.valid?
# before type check
# => nil

outcome = Example.run(string: '')
# before type check
# before validate
outcome.valid?
# before type check
# before validate
# => false

outcome = Example.run(string: 'invalid')
# before type check
# before validate
# before execute
outcome.valid?
# => false

outcome = Example.run(string: 'valid')
# before type check
# before validate
# before execute
outcome.valid?
# => true
```

Returning a different outcome class is tricky. Interactions are designed to be used as form objects. If the outcome was a different class, that class would end up looking a lot like an interaction.

**Comment by tfausak** [comment]
Are you satisfied with my explanation? Is there anything we can do to help you out here?

**Comment by guizmaii** [comment]
Sorry for the late response, I was in vacations.

Unfortunately, I'm not satisfied. 
I don't understand how you couldn't be shocked by the fact that the validation is call twice and thus I must put my "@callback_called_once trick" in order to execute only once my business logic.

If anyone want to do the same thing like me, i.e. execute something (once) when type checking fail, he will have to add the same "trick" !

Type checking is one of the great functionality of this gem. Execute a default behavior when the type check fail should be a functionnality too. 
More, in order to help users to execute controller specific behaviors, the outcome could contains  what failed during the ```run``` method execution (type check, validation check, execution).
Currently, in the controller, if I want to know if the type check has failed, I have to know which types of errors could be generated by type checking and look for them in the ```outcome.errors```. 
An ```outcome.type_check_passed?``` method could be simpler for users.

I will look at your code this week in order to do a pull request to show you what I'm talking about. Maybe, it'll be better for our discussion.

I'm sorry to bother you like this but your gem is, for me, the actual best interactor gem, it's only this little strange behavior which bother me.

**Comment by tfausak** [comment]
> I don't understand how you couldn't be shocked by the fact that the validation is call twice

I'm not at all surprised. In fact, I would be surprised if it didn't do that. [The `#valid?` documentation](http://api.rubyonrails.org/classes/ActiveModel/Validations.html#method-i-valid-3F) says:  "runs all the specified validations". It doesn't say, "runs all the specified validations *if they haven't yet been run*". In fact, the example code below that statement shows why this is desired behavior.

``` rb
person = Person.new
person.name = ''
person.valid? # => false
person.name = 'david'
person.valid? # => true
```

> the outcome could contains what failed during the `run` method execution

I don't think this is a direction we're going to move in. The difference between type checking and validation is only really important for the interaction, not the caller. If you need more information about how an interaction failed, you can already get it through the `errors` object. In particular, symbolic errors provide you a way to reliably pass information about failure from the interaction to the caller.

To make this discussion more concrete, can you describe what you're trying to do that can only be executed once? In particular, I'm curious about why it has to be in a callback instead of in `execute`.

**Comment by tfausak** [comment]
@guizmaii how can I help here?

**Comment by tfausak** [comment]
I'm going to close this issue because I think we have solved your original problem. We've also found a workaround for running validations exactly once. Please feel free to open another issue if you want to discuss moving away from how ActiveModel handles validation and callbacks.
